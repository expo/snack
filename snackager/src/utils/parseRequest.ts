import querystring from 'querystring';
import validate from 'validate-npm-package-name';

import { getCoreExternals } from '../bundler/externals';

export type BundleRequest = {
  qualified: string;
  scope?: string | null;
  id: string;
  tag: string;
  deep?: string | null;
  platforms: string[];
  rebuild: boolean;
  bypassCache: boolean;
  versionSnackager: boolean;
};

// parses a bundle request, resulting in these fields:
//
// example: `/@expo/configurator/submodule@^1.0.0?platforms=ios,android
//
// scope: '@expo'
// id:    'configurator'
// deep:  'submodule'
// tag:   '^1.0.0'
// platforms: ['ios', 'android', 'web]
//
// all are optional except the id
export default function parseRequest(url: string): BundleRequest {
  const match = /^\/(?:@([^/?]+)\/)?([^@/?]+)(?:\/([^@]+?))?(?:@(.+?))?(?:\?(.+))?$/.exec(url);

  if (!match) {
    throw new Error('Failed to parse request');
  }

  const scope = match[1]; // matches scope in `@scope/package`
  const id = match[2]; // matches id in `@user/id` or `package`
  const deep = match[3]; // matches deep path in `package/debounce` or `package/debounce@^3.4.0`
  const tag = match[4] || 'latest'; // matches version number in `package@^3.4.0`
  const qs = match[5]; // matches the query string

  if (!id) {
    throw new Error('Failed to determine the module name');
  }

  const validPackageResult = validate(id);
  const validPackage =
    validPackageResult.validForNewPackages ||
    !validPackageResult.warnings ||
    validPackageResult.warnings.every((warning) => warning.includes('is a core module name'));

  if (!validPackage) {
    throw new Error(`Module '${id}' is not a valid package name`);
  }

  const qualified = scope ? `@${scope}/${id}` : id;
  const query = qs ? querystring.parse(qs) : null;
  const deepQualified = `${qualified}${deep ? `/${deep}` : ''}`;

  if (getCoreExternals().includes(deepQualified)) {
    throw new Error(`Bundling core module '${deepQualified}' is prohibited`);
  }

  // TODO: check if query.platforms can be a `string[]`
  const platforms = query?.platforms
    ? (query.platforms as string).split(',').filter((p) => p)
    : ['ios', 'android', 'web'];

  return {
    qualified,
    scope,
    id,
    tag,
    deep,
    platforms,
    rebuild: Boolean(query && query.rebuild === 'true' && process.env.DEBUG_LOCAL_FILES),
    bypassCache: Boolean(query && query.bypassCache === 'true'),
    versionSnackager: Boolean(query && query.version_snackager === 'true'),
  };
}
