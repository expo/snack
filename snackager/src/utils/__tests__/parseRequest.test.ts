import parseRequest from '../parseRequest';

it('gets id from request', () => {
  const parsed = parseRequest('/lodash');
  expect(parsed).toMatchSnapshot();
});

it('gets user from request', () => {
  const parsed = parseRequest('/@scoped/lodash');
  expect(parsed).toMatchSnapshot();
});

it('gets tag from request', () => {
  const parsed = parseRequest('/lodash@14.5.2');
  expect(parsed).toMatchSnapshot();
});

it('gets deep path from request', () => {
  const parsed = parseRequest('/lodash/bounce/index');
  expect(parsed).toMatchSnapshot();
});

it('gets deep path and tag from request', () => {
  const parsed = parseRequest('/lodash/bounce/index@14.5.2');
  expect(parsed).toMatchSnapshot();
});

it('gets tag from request', () => {
  const parsed = parseRequest('/lodash/bounce/index?platforms=ios');
  expect(parsed).toMatchSnapshot();
});

it('gets platforms from request', () => {
  const parsed = parseRequest('/lodash/bounce/index?platforms=ios,android');
  expect(parsed).toMatchSnapshot();
});

it('gets all fields from request', () => {
  const parsed = parseRequest('/@scoped/lodash/bounce/index@14.5.2?platforms=ios');
  expect(parsed).toMatchSnapshot();
});
