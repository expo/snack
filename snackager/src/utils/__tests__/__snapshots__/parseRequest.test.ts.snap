// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets all fields from request 1`] = `
Object {
  "bypassCache": false,
  "deep": "bounce/index",
  "id": "lodash",
  "platforms": Array [
    "ios",
  ],
  "qualified": "@scoped/lodash",
  "rebuild": false,
  "scope": "scoped",
  "tag": "14.5.2",
  "versionSnackager": false,
}
`;

exports[`gets deep path and tag from request 1`] = `
Object {
  "bypassCache": false,
  "deep": "bounce/index",
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "14.5.2",
  "versionSnackager": false,
}
`;

exports[`gets deep path from request 1`] = `
Object {
  "bypassCache": false,
  "deep": "bounce/index",
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "latest",
  "versionSnackager": false,
}
`;

exports[`gets id from request 1`] = `
Object {
  "bypassCache": false,
  "deep": undefined,
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "latest",
  "versionSnackager": false,
}
`;

exports[`gets platforms from request 1`] = `
Object {
  "bypassCache": false,
  "deep": "bounce/index",
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "latest",
  "versionSnackager": false,
}
`;

exports[`gets tag from request 1`] = `
Object {
  "bypassCache": false,
  "deep": undefined,
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "14.5.2",
  "versionSnackager": false,
}
`;

exports[`gets tag from request 2`] = `
Object {
  "bypassCache": false,
  "deep": "bounce/index",
  "id": "lodash",
  "platforms": Array [
    "ios",
  ],
  "qualified": "lodash",
  "rebuild": false,
  "scope": undefined,
  "tag": "latest",
  "versionSnackager": false,
}
`;

exports[`gets user from request 1`] = `
Object {
  "bypassCache": false,
  "deep": undefined,
  "id": "lodash",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "qualified": "@scoped/lodash",
  "rebuild": false,
  "scope": "scoped",
  "tag": "latest",
  "versionSnackager": false,
}
`;
